### ISSUE TITLE
eirini-loggregator-bridge is missing logs

### ISSUE THREAD

**Issue by mudler** [SUSE/eirini-loggregator-bridge#6]
This is describing the issue found while testing https://github.com/cloudfoundry-community/eirini-bosh-release/pull/74 . Since this seems to be a separate issue, I'm documenting the findings here.

I've pushed the sample-ticking app, and scaled to 10 instances:

```
   2020-03-24T12:27:06.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:06                                                                 [0/0]
   2020-03-24T12:27:07.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:07                                                                     
   2020-03-24T12:27:07.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:07                                                                     
   2020-03-24T12:27:08.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:08                                                                     
   2020-03-24T12:27:08.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:08                                                                     
   2020-03-24T12:27:09.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:09                                                                     
   2020-03-24T12:27:09.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:09                                                                     
   2020-03-24T12:27:10.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:10                                                                     
   2020-03-24T12:27:10.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:10                                                                     
   2020-03-24T12:27:11.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:11                                                                     
   2020-03-24T12:27:11.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:11                                                                     
   2020-03-24T12:27:12.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:12                                                                     
   2020-03-24T12:27:12.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:12                                                                     
   2020-03-24T12:27:13.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:13                                                                     
   2020-03-24T12:27:13.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:13                                                                     
   2020-03-24T12:27:14.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:14                                                                     
   2020-03-24T12:27:14.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:14                                                                     
   2020-03-24T12:27:15.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:15                                                                     
   2020-03-24T12:27:15.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:15                                                                     
   2020-03-24T12:27:16.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:16                                                                     
   2020-03-24T12:27:16.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:16                                                                     
   2020-03-24T12:27:17.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:17                                                                     
   2020-03-24T12:27:17.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:17                                                                     
   2020-03-24T12:27:17.00+0000 [APP/PROC/WEB/3] OUT [ab874e81-367a-4645-a87b-5618de5a9ca5:3] Ticking 2020-03-24 12:27:17                                                                     
   2020-03-24T12:27:18.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:18                                                                     
   2020-03-24T12:27:18.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:18                                                                     
   2020-03-24T12:27:19.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:19                                                                     
   2020-03-24T12:27:19.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:19                                                                     
   2020-03-24T12:27:20.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:20                                                                     
   2020-03-24T12:27:20.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:20                                                                     
   2020-03-24T12:27:21.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:21                                                                     
   2020-03-24T12:27:21.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:21                                                                     
   2020-03-24T12:27:22.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:22
   2020-03-24T12:27:22.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:22
   2020-03-24T12:27:23.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:23
   2020-03-24T12:27:23.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:23
   2020-03-24T12:27:24.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:24
   2020-03-24T12:27:24.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:24
   2020-03-24T12:27:25.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:27:25
   2020-03-24T12:27:25.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:27:25
   2020-03-24T12:27:25.00+0000 [APP/PROC/WEB/3] OUT [ab874e81-367a-4645-a87b-5618de5a9ca5:3] Ticking 2020-03-24 12:27:25
   2020-03-24T12:27:26.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:27:26
   2020-03-24T12:27:26.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:27:26
```

As you can see, the instance 3 got less messages in, but at the same time, each second we get at maximum 2/3 batches of messages, even if we should see 10 instances. 

The eirini pod got no restarts so far:
```
susecf-scf-eirini-0                    9/9     Running   0          55m
```
So it seems we ignore watcher timeout errors, but we don't dispatch all the messages. To support this:

```
 2020-03-24T12:31:45.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:31:45                                                                     
   2020-03-24T12:31:45.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:31:45                                                                     
   2020-03-24T12:31:45.00+0000 [APP/PROC/WEB/3] OUT [ab874e81-367a-4645-a87b-5618de5a9ca5:3] Ticking 2020-03-24 12:31:45                                                                     
   2020-03-24T12:31:45.00+0000 [APP/PROC/WEB/1] OUT [dfc607c5-88fe-4305-a3dc-9dc6bc241613:1] Ticking 2020-03-24 12:31:45                                                                     
   2020-03-24T12:31:45.00+0000 [APP/PROC/WEB/8] OUT [57d26343-354b-444b-a748-d55997a4683c:8] Ticking 2020-03-24 12:31:45                                                                     
   2020-03-24T12:31:46.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:31:46                                                                     
   2020-03-24T12:31:46.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:31:46                                                                     
   2020-03-24T12:31:47.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:31:47                                                                     
   2020-03-24T12:31:47.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:31:47                                                                     
   2020-03-24T12:31:48.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:31:48                                                                     
   2020-03-24T12:31:48.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:31:48                                                                     
   2020-03-24T12:31:49.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:31:49                                                                     
   2020-03-24T12:31:49.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:31:49                                                                     
   2020-03-24T12:31:50.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:31:50                                                                     
   2020-03-24T12:31:50.00+0000 [APP/PROC/WEB/6] OUT [0e7a5668-2a3c-4650-ae06-737b6d42a877:6] Ticking 2020-03-24 12:31:50                                                                     
   2020-03-24T12:31:51.00+0000 [APP/PROC/WEB/4] OUT [267fd49a-a829-448f-b5fd-00387263e940:4] Ticking 2020-03-24 12:31:51                                                                     
   2020-03-24T12:31:51.00+0000 [APP/PROC/WEB/5] OUT [bca85f4d-a745-4539-a13f-297136c703fe:5] Ticking 2020-03-24 12:31:51                                                                     
   2020-03-24T12:31:52.00+0000 [APP/PROC/WEB/0] OUT [695cfbda-e685-415d-b304-8a43f0877904:0] Ticking 2020-03-24 12:31:52      
 ```

As you can see, now we got messages from instance 8, but only once. It looks like they are racing for "slots" (that we don't have).

I've tried lowering the watcher channels, 

```
sudo sysctl fs.inotify.max_user_instances=20
sudo sysctl fs.inotify.max_user_watches=20
```

to exclude that is related to kernel limits. in the loggregator bridge I can observe: 

```
failed to create fsnotify watcher: too many open files
``` 

I can see this message only when I lower the limit, as far as I can tell that's not the root cause.

Now the bridge doesn't crash anymore, and I can see no pod restarts:
```
susecf-scf-eirini-0                    9/9     Running   0          71m
```

But as before we were restarting and re-streaming all the logs, now we probably miss the logs when the watcher timeouts. I can't see why this happens as I would have expected messages to be delayed, but not just "vanished"

**Comment by mudler** [comment]
Can't reproduce this anymore, I've left a cluster running overnight with a ticking up scaled to 20. In the previous reports, 10 instances were enough to show the issue.

As I couldn't observe this behavior anymore, I've created a small script to keep track of the log instances and count them:

```
$VAR1 = {
          '[APP/PROC/WEB/10]' => 24878,
          '[APP/PROC/WEB/8]' => 24878,
          '[APP/PROC/WEB/18]' => 24879,
          '[APP/PROC/WEB/2]' => 24879,
          '[APP/PROC/WEB/5]' => 24879,
          '[APP/PROC/WEB/7]' => 24879,
          '[APP/PROC/WEB/3]' => 24879,
          '[APP/PROC/WEB/17]' => 24879,
          '[APP/PROC/WEB/11]' => 24879,
          '[APP/PROC/WEB/6]' => 24879,
          '[APP/PROC/WEB/9]' => 24878,
          '[APP/PROC/WEB/13]' => 24879,
          '[APP/PROC/WEB/15]' => 24878,
          '[APP/PROC/WEB/1]' => 24879,
          '[APP/PROC/WEB/0]' => 24879,
          '[APP/PROC/WEB/16]' => 24878,
          '[APP/PROC/WEB/14]' => 24878,
          '[APP/PROC/WEB/4]' => 24879,
          '[APP/PROC/WEB/12]' => 24879,
          '[APP/PROC/WEB/19]' => 24879,
        };
```

20 instances, overnight produced ~24k lines of logs ( to note, the +1 delta is due to discrepancy when the script started). 

I've tried again bumping the instances to 40, but without luck:

```
$VAR1 = { 
          '[APP/PROC/WEB/20]' => 2351,                                                                                                                                       
          '[APP/PROC/WEB/15]' => 2351, 
          '[APP/PROC/WEB/16]' => 2351,                                                                                                                              
          '[APP/PROC/WEB/12]' => 2351,
          '[APP/PROC/WEB/1]' => 2351, 
          '[APP/PROC/WEB/0]' => 2351, 
          '[APP/PROC/WEB/30]' => 2351,
          '[APP/PROC/WEB/18]' => 2351,
          '[APP/PROC/WEB/34]' => 2351,
          '[APP/PROC/WEB/27]' => 2351,
          '[APP/PROC/WEB/23]' => 2351,
          '[APP/PROC/WEB/39]' => 2351,
          '[APP/PROC/WEB/33]' => 2351,
          '[APP/PROC/WEB/29]' => 2351,
          '[APP/PROC/WEB/24]' => 2351,
          '[APP/PROC/WEB/37]' => 2351,
          '[APP/PROC/WEB/11]' => 2351,
          '[APP/PROC/WEB/14]' => 2351,
          '[APP/PROC/WEB/21]' => 2351,
          '[APP/PROC/WEB/4]' => 2351,
          '[APP/PROC/WEB/9]' => 2351, 
          '[APP/PROC/WEB/19]' => 2351,
          '[APP/PROC/WEB/13]' => 2351,
          '[APP/PROC/WEB/8]' => 2351, 
          '[APP/PROC/WEB/7]' => 2351, 
          '[APP/PROC/WEB/17]' => 2351,
          '[APP/PROC/WEB/31]' => 2351,
          '[APP/PROC/WEB/6]' => 2351,
          '[APP/PROC/WEB/28]' => 2351,
          '[APP/PROC/WEB/35]' => 2351,
          '[APP/PROC/WEB/36]' => 2351,
          '[APP/PROC/WEB/2]' => 2351, 
          '[APP/PROC/WEB/5]' => 2351, 
          '[APP/PROC/WEB/32]' => 2351,
          '[APP/PROC/WEB/22]' => 2351,
          '[APP/PROC/WEB/3]' => 2351,
          '[APP/PROC/WEB/10]' => 2351,
          '[APP/PROC/WEB/26]' => 2351,
          '[APP/PROC/WEB/25]' => 2351,
          '[APP/PROC/WEB/38]' => 2351 
        };          
``` 

Here is the script being used to summarize the received messages:

```perl
#!/usr/bin/perl
use Data::Dumper;
open(TICKING, "cf logs ticking |");
my $pids;
while (<TICKING>) {
  ($date, $pid, undef, undef, $message) = split;
  next if $pid eq "" || $pid eq "logs";
  $pids->{$pid}++;
  print Dumper($pids);
}

close(TICKING);
```

**Comment by mudler** [comment]
```
 cf apps             
Getting apps in org system / space tmp as admin...
OK

name        requested state   instances   memory   disk   urls
ticking     started           40/40       20M      20M    ticking.172.17.0.3.omg.howdoi.website
cf-nodejs   started           1/1         512M     1G     cf-nodejs-spontaneous-bat-hr.172.17.0.3.omg.howdoi.website
```

A second round of logs, shows still no issues:
```
$VAR1 = {                                                                                                                                                                                     
          '[APP/PROC/WEB/14]' => 2911,                                                                                                                                                                  '[APP/PROC/WEB/19]' => 2911,
          '[APP/PROC/WEB/35]' => 2911,
          '[APP/PROC/WEB/6]' => 2911,                                                                                                                                                                   '[APP/PROC/WEB/27]' => 2911,
          '[APP/PROC/WEB/32]' => 2911,
          '[APP/PROC/WEB/26]' => 2911,                                                                                                                                                                  '[APP/PROC/WEB/0]' => 2911,
          '[APP/PROC/WEB/10]' => 2911,
          '[APP/PROC/WEB/31]' => 2911,                                                                                                                                                                  '[APP/PROC/WEB/29]' => 2911,
          '[APP/PROC/WEB/24]' => 2911,
          '[APP/PROC/WEB/38]' => 2911,                                                                                                                                                                  '[APP/PROC/WEB/20]' => 2911,
          '[APP/PROC/WEB/3]' => 2911,
          '[APP/PROC/WEB/33]' => 2911,
          '[APP/PROC/WEB/7]' => 2911,
          '[APP/PROC/WEB/17]' => 2911,      
          '[APP/PROC/WEB/16]' => 2911,
          '[APP/PROC/WEB/9]' => 2911,
          '[APP/PROC/WEB/21]' => 2911,
          '[APP/PROC/WEB/1]' => 2911,        
          '[APP/PROC/WEB/18]' => 2911,
          '[APP/PROC/WEB/36]' => 2911,
          '[APP/PROC/WEB/13]' => 2911,
          '[APP/PROC/WEB/37]' => 2911,
          '[APP/PROC/WEB/22]' => 2911,
          '[APP/PROC/WEB/25]' => 2911, 
          '[APP/PROC/WEB/5]' => 2911,                                                          
          '[APP/PROC/WEB/23]' => 2911,
          '[APP/PROC/WEB/4]' => 2911,                                                                                                                                                         
          '[APP/PROC/WEB/12]' => 2911,                                                                                                                                                        
          '[APP/PROC/WEB/8]' => 2911,
          '[APP/PROC/WEB/2]' => 2911,
          '[APP/PROC/WEB/11]' => 2911,
          '[APP/PROC/WEB/30]' => 2911,
          '[APP/PROC/WEB/28]' => 2911,
          '[APP/PROC/WEB/34]' => 2911,
          '[APP/PROC/WEB/39]' => 2911,
          '[APP/PROC/WEB/15]' => 2911
        };
```

**Comment by mudler** [comment]
This could explain the issue I was experiencing originally: https://github.com/kubernetes/client-go/commit/33c1f9f2a43292d0b2dcb00f435bf3becad0528d#diff-b414e368c7101665eb44ed4a6cba77ef

**Comment by mudler** [comment]
I've created https://github.com/SUSE/eirinix/issues/23 as it's not so straightforward right now to bump kubernetes/client-go as we depend on cf-operator in EiriniX.
